src/udev/udevadm-trigger.c

```
@@ -270,18 +289,87 @@ static int adm_trigger(struct udev *udev, int argc, char *argv[]) {
-                return 0;
+                break;
         case TYPE_DEVICES:
                 udev_enumerate_scan_devices(udev_enumerate);
-                exec_list(udev_enumerate, action);
-                return 0;
+                break;
         default:
                 assert_not_reached("device_type");
         }
+        exec_list(udev_enumerate, action, settle ? settle_set : NULL);
+
+        if (settle) {
+                while (!set_isempty(settle_set)) {
+                        int fdcount;
+                        struct epoll_event ev[4];
+                        int i;
+
+                        fdcount = epoll_wait(fd_ep, ev, ELEMENTSOF(ev), -1);
+                        if (fdcount < 0) {
+                                if (errno != EINTR)
+                                        fprintf(stderr, "error receiving uevent message: %m\n");
+                                continue;
+                        }
+
+                        for (i = 0; i < fdcount; i++) {
+                                if (ev[i].data.fd == fd_udev && ev[i].events & EPOLLIN) {
+                                        struct udev_device *device;
+                                        const char *syspath = NULL;
+
+                                        device = udev_monitor_receive_device(udev_monitor);
+                                        if (device == NULL)
+                                                continue;
+                                        syspath = udev_device_get_syspath(device);
+                                        if (verbose)
+                                                printf("settle %s\n", syspath);
+                                        if (set_contains(settle_set, syspath)) {
+                                                set_remove(settle_set, syspath);
+                                        } else {
+                                                fprintf(stderr, "syspath set does not contain '%s'\n", syspath);
```

That's not really informative for the caller of the binary. They wouldn't know what went wrong anyway. Maybe you can just ignore the error handling here, and just emit a debug line in that case. Also, there's no need to check if the set contains the path before removal. Something like
```C
     if (!set_remove(settle_set, syspath))
         log_debug("Got epoll event on syspath %s not present in syspath set", syspath);
```